### ISSUE TITLE
Error writing an endpoint

### ISSUE THREAD

**Issue by gvlpedro** [zio/zio-http#2535]
**Describe the bug**
I am using the version 3.0.0-RC3 and I had a cryptic error writing a simple endpoint
**To Reproduce**
Here the code I wrote (let me know if I am writing something in a wrong way). After reducing the code I got the wrong line:

```
object DomainController {
  def apply(): HttpApp[Any] = {
    Routes(
      Method.POST / string("id") / "domain" / "create" ->
        handler { (id: String, req: Request) =>
          for {
            //_ <- ZIO.logInfo("URL:" + req.url.path.toString)
            out <-ZIO.succeed(Response.text(s"Created domain '$id'."))
            _ <- req.body.asString.map(Response.text(_)) // TODO: There is a bug in this line
          } yield out
        }
    ).toHttpApp @@ ComposedMiddlewares()
  }
}
```
I created a [github project ](https://github.com/gvlpedro/myzio/blob/907fe78d633f02f9156f730cff050db6f44816fb/src/main/scala/com/example/myzio/v1/controller/DomainController.scala#L25) to help you to reproduce the error.

When I comment that line all works correctly, however if this line is there the project fails in a non specific error.

**Error**
**scala: ## Exception when compiling 3 sources to /Users/user/myzio/target/scala-3.3.1/classes
java.lang.AssertionError: assertion failed
scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
dotty.tools.dotc.ast.tpd$.TypeApply(tpd.scala:60)
dotty.tools.dotc.ast.tpd$TreeOps$.appliedToTypeTrees$extension(tpd.scala:985)
...**

**Desktop (please complete the following information):**
 MacOS

Am I doing something wrong? or it is a real bug?

**Comment by onsoup42** [comment]
I ran in to the same issue. The below code will make the Scala compiler crash. 
```scala
  val response: ZIO[Any, Throwable, Response] = ZIO.fail(???)

  val app: HttpApp[Client] =
    Routes(
      Method.GET / "test" -> handler(response)
    ).toHttpApp
```

The actual issue is that the route doesn't handle it's errors. So the fix is quite simple, handle the errors:

```scala
  val response: ZIO[Any, Throwable, Response] = ZIO.fail(???)

  val app: HttpApp[Client] =
    Routes(
      Method.GET / "test" -> handler(response).mapError(error => Response.internalServerError(error.getMessage))
    ).toHttpApp
```

However, I believe this is still bug in zio-http. This should produce a type error, not a compile crash.

I tried this on the latest snapshot
`"dev.zio" %% "zio-http" % "3.0.0-RC4+27-f2ce5cfd-SNAPSHOT"`
Using Scala 3.3.1 running on MacOS both using AppleSilicon and x86

**Comment by 987Nabil** [comment]
Scala compiler bug. Will be fixed with 3.3.2/3.4
https://github.com/lampepfl/dotty/issues/18650
